### PR TITLE
introduce -Submit my Revisions- for author and new status for reviewRound

### PR DESCRIPTION
**Pull request by forgive38**
Hi,

this PR is an option for #2618

it's also a response to: https://forum.pkp.sfu.ca/t/what-is-request-revisions-or-resubmit-for-review/33702
and https://forum.pkp.sfu.ca/t/notification-of-revision-upload/31676/4

Because we have found that the feed for the author after revisions have been requested not so clear, especially for "Submission has been resubmitted for review", I propose to introduce a new status: REVIEW_ROUND_STATUS_SUBMISSION_RESUBMITTED

and disable the detection state
https://github.com/forgive38/pkp-lib/blob/e20dddc7dbfbd01836af39a231172ab94db9894d/classes/submission/reviewRound/ReviewRound.inc.php#L135

Then I propose a new action in AuthorReviewRevisionsGridHandler.inc.php for the author that is "Inform that revisions are available", which could be "Send my revisions".

sendrevisions

thank you for reading this PR : I'm open to comments and better way to do ...

### CODE REVIEW

File: classes/notification/managerDelegate/EditorDecisionNotificationManager.inc.php [hunk 1]
```diff
@@ -154,6 +154,7 @@ function _getAllEditorDecisionNotificationTypes() {
 	function _getNotificationTaskLevel($type) {
 		switch ($type) {
 			case NOTIFICATION_TYPE_EDITOR_DECISION_PENDING_REVISIONS:
+                        case NOTIFICATION_TYPE_EDITOR_DECISION_RESUBMIT:
```
Comment (asmecher) [inline classes/notification/managerDelegate/EditorDecisionNotificationManager.inc.php]: Make sure your editor is set to indent with tabs, not spaces, thanks!

File: classes/submission/reviewRound/ReviewRound.inc.php [hunk 2]
```diff
@@ -130,23 +132,26 @@ public function determineStatus() {
...
-		// submitted
-		if ($this->getStatus() == REVIEW_ROUND_STATUS_REVISIONS_REQUESTED || $this->getStatus() == REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED) {
-			$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
-			$submissionFiles = $submissionFileDao->getRevisionsByReviewRound($this, SUBMISSION_FILE_REVIEW_REVISION);
-			if (empty($submissionFiles)) {
-				return REVIEW_ROUND_STATUS_REVISIONS_REQUESTED;
-			}
-			return REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED;
-		}
+//		// If revisions have been requested, check to see if any have been
```
Comment (asmecher) [inline classes/submission/reviewRound/ReviewRound.inc.php]: Feel free to remove unnecessary code if you've implemented this a different way. Commented code clutters up the repo.

File: locale/en_US/common.xml [hunk 3]
```diff
@@ -288,6 +288,8 @@
 	<message key="debug.notes">Notes:</message>
 	<message key="defaultMetric.title">Default Metric</message>
 	<message key="defaultMetric.availableMetrics">Available Metrics</message>
+        <message key="editor.submissionReview.sendRevisions">Inform that revisions are available</message>
+        <message key="editor.submissionReview.confirmSendRevisions">Inform that revisions are available ?</message>
```
Comment (asmecher) [inline locale/en_US/common.xml]: Remove the space before the `?`

File: locale/en_US/submission.xml [hunk 4]
```diff
@@ -157,6 +157,93 @@
 	<message key="submission.changeComments">Change to</message>
 	<message key="submission.changeFile">Change File</message>
 	<message key="submission.citations">References</message>
+	<message key="submission.citations.editor.citationlist.approved">Approved</message>
```
Comment (asmecher) [inline locale/en_US/submission.xml]: Are these intentional additions? I don't think they're relevant to this change; perhaps an accidental revert got added to the PR.

**Comment by forgive38** [comment]
Sorry for delay
I do the changes requested

Thank you

File: classes/submission/reviewRound/ReviewRound.inc.php [hunk 5]
```diff
@@ -130,23 +132,13 @@ public function determineStatus() {
...
-
-		// Leave the status alone if it is set to one of the EditorDecisions
-		// which have advanced the submission beyond this round
 		$statusFinished = in_array(
 			$this->getStatus(),
 			array(
+                                REVIEW_ROUND_STATUS_REVISIONS_REQUESTED,
+                                REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED,
 				REVIEW_ROUND_STATUS_RESUBMITTED,
+                                REVIEW_ROUND_STATUS_SUBMISSION_RESUBMITTED,
```
Comment (asmecher) [inline classes/submission/reviewRound/ReviewRound.inc.php]: Watch out for space-based indents -- they should all be tabs.

File: controllers/grid/files/review/AuthorReviewRevisionsGridHandler.inc.php [hunk 6]
```diff
@@ -29,11 +29,97 @@ function __construct() {
...
 		$this->addRoleAssignment(
 			array(ROLE_ID_AUTHOR),
-			array('fetchGrid', 'fetchRow')
+			array('fetchGrid', 'fetchRow','sendRevisions')
 		);
 
 		$this->setTitle('editor.submission.revisions');
 	}
+        
+        function initialize($request) {
```
Comment (asmecher) [inline controllers/grid/files/review/AuthorReviewRevisionsGridHandler.inc.php]: Make sure you include a header on all functions -- even if it's something brief like...

    /**
     * @copydoc GridHandler::initialize()
     */

File: locale/en_US/manager.xml [hunk 7]
```diff
@@ -515,7 +559,7 @@
 	<message key="manager.navigationMenus.customPage">Custom Page</message>
 	<message key="manager.navigationMenus.customPage.description">Create a custom page on your site and link to it from a navigation menu.</message>
 	<message key="manager.navigationMenus.remoteUrl">Remote URL</message>
-	<message key="manager.navigationMenus.remoteUrl.description">Link to any URL on another site, like https://pkp.sfu.ca.</message>
```
Comment (asmecher) [inline locale/en_US/manager.xml]: It looks like you need to rebase your pull request -- this reverts changes.

**Comment by forgive38** [comment]
I do a cleaner commit, I hope...

**Comment by forgive38** [comment]
Maybe this PR can be an answer to #3326 and #3278

**Comment by NateWr** [comment]
Thanks for the PR @forgive38! This is a UX issue we've struggled with for a while. I like your solution of providing a final confirmation button. 

However, I'm a bit worried that it's not very clear to the author they need to signal when all of their revisions are done. The button is in an obvious place, but a lot of users just don't go looking for something unless they know they need it. For this to work, we need a better way to signal to the author that they are expected to confirm their revision files are available.

My thinking has been that we'll need a more "wizard"-like workflow for this, where the author uploads files in a multi-step process, with the final step being a "submit revisions to editor" step. But that's a big overhaul of the system we have.

If we decided to go ahead with a temporary solution like this, we'll need to think of some ways to signal to the author what's expected of them. A couple of ideas:

- In the notification box at the top, we could include a link at the top, which opened a confirmation window. Something like: "New revisions have been requested. Once you have uploaded them below, [submit them to the editor for review](#)". Then the link would open a confirmation window showing what files were to be submitted.
- Perhaps after a single revision file has been uploaded, the notification box could change to: "You've submitted X revision files. Is that everything? [Submit revisions](#)".

**Comment by forgive38** [comment]
@NateWr : I see what you mean...

But for the moment, this operation is enough for our users: the status of the submission does not change as long as the author does not press the button. And so if he does not validate his revisions, he will see that in the list of submissions the status is still at "Waiting for revisions"

**Comment by bozana** [comment]
I can try this PR and see if I can somehow integrate the proposals from @NateWr. @NateWr, would that be the preferred solution for now (3.1.1)? Or something else? Also, what would be the preferred minimal solution for now? -- I can try to implement the preferred solution, and then maybe to also have a minimal solution that we can take, if I do not manage to implement that preferred solution :-)

A minimal solution could be to:
-- add one review round status for revisions after the decision "resubmit for review" and provide the appropriate review round status notification (as it is solved in this PR)
-- send automatically an e-mail every time a revision is uploaded. This will introduce the e-mail template again, but... maybe the overall solution in the future requires it however?
This would however not give the possibility to the author to notify the editor about the revisions uploaded when he/she wants...

THANKS!!!

**Comment by NateWr** [comment]
@bozana that might work. I don't think I know the details well enough here to say exactly what would or wouldn't work as a minimum implementation.

For example, what does `REVIEW_ROUND_STATUS_SUBMISSION_RESUBMITTED` add that we aren't already signalling with `REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED`? If we send an email with each uploaded revision, we don't need any "indicate revisions complete" button, do we?

Improving the notification messages is probably the best minimum implementation, so that it's clearer to authors what they need to do and clear to editors when revisions have been submitted. Beyond that, I'd prefer to do nothing unless we're implementing more of a wizard-like flow for authors.

**Comment by bozana** [comment]
@forgive38, to be 100% sure I understand the code correctly: if the author clicks on "inform that revisions are available" first then the review round status will be changed and then the editor will see the notification on that review round page, correct? -- No e-mail will be sent to editor, correct?

**Comment by bozana** [comment]
@NateWr, I believe it would be easier/clearer to have an extra review round status for when the revisions are uploaded after "resubmit for review" decision/status: S. for example this code: https://github.com/pkp/pkp-lib/blob/master/classes/submission/reviewRound/ReviewRound.inc.php#L135-L142. If we would use the same status for any revision upload (no matter what was the decision/status before) and if the revisions would then be deleted, we would not know what should be the review round status again i.e. we would need to see what was the last editor decision for that round to match the review round status. This would work too, but maybe it is easier and clearer to have a new review round status so that we exactly and immediately know what it means. What do you think?

Else, then the question is how to solve the notification for now: 
a) as suggested in this PR, if I understand it correctly: if the author clicks on "inform that revisions are available" first then the review round status will be changed and then the editor will see the notification on that review round page
b) to change the review status (and then the editor will see the notification on that review round page) with a first revision upload -- as it is now, just with the additional status change when the revision is uploaded after "resubmit for review"
c) could add an e-mail notification to a) or b) about the status change (as it was earlier, I believe) -- no e-mail template needed, I believe
d) could add an automatic e-mail notification to b) every time a revision is uploaded -- an e-mail template needed
e) could be a combination of a) and b) -- to change the status with the first revision upload and to use the button for an e-mail notification
or eventually any combination of these solutions :-)

**Comment by NateWr** [comment]
@bozana Hmm, I'm not sure I'm plugged in enough on this to say one way or the other about the review round status. If it needs a new status I trust you on that one.

I think I'd prefer to avoid a button altogether until we can go to a wizard-like revision submission process. I'm just worried about a "hung" revisions process where the author has uploaded revisions but failed to click the "complete" button.

That would mean (b) and (c), I think, from your list above. But if it provides useful notifications, however you do it is good for me, as long as we avoid the possibility of a "hung" revision submission.

**Comment by forgive38** [comment]
@NateWr : 

> For example, what does REVIEW_ROUND_STATUS_SUBMISSION_RESUBMITTED add that we aren't already signalling with REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED? If we send an email with each uploaded revision, we don't need any "indicate revisions complete" button, do we?

Ok, I see my mistake: I created a new status 'REVIEW_ROUND_STATUS_RESUBMITTED' which has not the right name, it should be "REVIEW_ROUND_STATUS_MAJOR_REVISIONS_SUBMITTED" and https://github.com/forgive38/pkp-lib/blob/f0f2e6cffc399b2b804b0b875452edbb1079f373/classes/submission/reviewRound/ReviewRound.inc.php#L22 should be rename 'REVIEW_ROUND_STATUS_MAJOR_REVISIONS_REQUESTED'

'REVIEW_ROUND_STATUS_MAJOR_REVISIONS_SUBMITTED' is not mandatory, maybe 'REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED' is enough, but we need something to inform editor that is a major revision.

@bozana : correct, after author clicks on the button, no mail are sent. For us, using the dashboard, which is clear, is enough.

For me, a) and c) can be a good solution or a wizard like submission process

**Comment by bozana** [comment]
OK, I think we agreed on/can accept the new review round status, as well as that an e-mail notification can be sent when that review round status changes.
The only thing we do not totally agree on is when the review round status should be changed: The PR suggests the change first when the button is clicked. But for now we would actually like to change the status immediately i.e. not to leave it in author's responsibility, because we are not sure he/she would exactly know what he/she would need to do. @forgive38, why wouldn't you want to have it immediately, after the first revision file is uploaded?
Thus, I would suggest that I take this PR as basis and adapt it for our current needs for 3.1.1: We would have the new review round status (and the new task "resubmit for review" from the PR). I would remove the button and instead keep the status change on the first revision file upload (this time for both, "revisions requested" and "resubmit for review") and add the e-mail notification for that status change.
The only thing left would be to give the author possibility to inform the editor about an additional revision file upload (that would not change the review round status), but currently this could be done with discussions or out of the system.
I think the authors and editors would then actually sufficiently be informed, with the status change on the review round page and e-mail notification (probably only to the editor?).
Later we can see how to improve it even more, with the wizard-like solution.

@NateWr, the status change (to the current REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED) is already happening correctly if the last "confirm" step is omitted for some reason -- thus this should be no problem also for the new status change after the "resubmit for review".

And one final question: When I am doing those changes, would you like me to change that review status constant name REVIEW_ROUND_STATUS_RESUBMITTED that seems to be confusing?
So maybe to double check with you what text shall we use for all that: 
For "revisions requested" we have now:
Task = Revision required.
REVIEW_ROUND_STATUS_REVISIONS_REQUESTED = Revisions have been requested.
REVIEW_ROUND_STATUS_REVISIONS_SUBMITTED = Revisions have been submitted.
For "resubmit for review" we would have:
Task = Resubmit for review.
And then the problematic ones:
REVIEW_ROUND_STATUS_RESUBMITTED = editor.submission.roundStatus.resubmitted = The submission must be resubmitted for another review round. (s. https://github.com/pkp/pkp-lib/issues/3326). Shall this be changed to REVIEW_ROUND_STATUS_MAJOR_REVISIONS_REQUESTED or REVIEW_ROUND_STATUS_RESUBMIT_REVISIONS_REQUESTED or ... ?
Is that text message OK?
And what would be the new status constant name and text message? REVIEW_ROUND_STATUS_MAJOR_REVISIONS_SUBMITTED or REVIEW_ROUND_STATUS_RESUBMIT_REVISIONS_SUBMITTED or REVIEW_ROUND_STATUS_REVISIONS_RESUBMITTED or ...?

Do you all agree @NateWr, @asmecher, @forgive38 ? :-) -- Please say yes :-DDD ;-)

**Comment by NateWr** [comment]
I would say:

`REVIEW_ROUND_STATUS_RESUBMITTED` -> `REVIEW_ROUND_STATUS_RESUBMIT_FOR_REVIEW`

Followed by:

`REVIEW_ROUND_STATUS_RESUBMIT_FOR_REVIEW_SUBMITTED`

File: locale/en_US/common.xml [hunk 8]
```diff
@@ -288,6 +288,8 @@
 	<message key="debug.notes">Notes:</message>
 	<message key="defaultMetric.title">Default Metric</message>
 	<message key="defaultMetric.availableMetrics">Available Metrics</message>
+        <message key="editor.submissionReview.sendRevisions">Inform that revisions are available</message>
```
Comment (asmecher) [inline locale/en_US/common.xml]: Watch out for space-based indents -- we use tabs.

File: locale/fr_FR/submission.xml [hunk 9]
```diff
@@ -334,7 +334,7 @@
 	<message key="notification.type.editorDecisionTitle">Dernière décision du rédacteur en chef.</message>
 	<message key="notification.type.editorDecisionAccept">Soumission acceptée.</message>
 	<message key="notification.type.editorDecisionExternalReview">Le processus d'évaluation externe a débuté.</message>
-	<message key="notification.type.editorDecisionPendingRevisions">Demande d'évaluation.</message>
```
Comment (asmecher) [inline locale/fr_FR/submission.xml]: Could you move these changes to a separate pull request? They aren't related to the issue this PR is associated with.

**Comment by bozana** [comment]
I am just looking how those notifications function: It seems that a notification can be either only displayed or displayed and an e-mail sent, but there is no possibility to just send an e-mail. Also the review round status notification function differently -- it is always displayed (the current review round status) and it does not have an user assigned, thus no e-mail will be sent. Thus I would need to create a new notification with the editor user assigned and this would lead to two similar notifications being displayed (the current review round status + the new notification that should send the e-mail).
I will investigate a little bit further, but if it is so we will have to use an e-mail template to inform the editor...

**Comment by bozana** [comment]
@forgive38, I will use your branch/changes from this PR as basis to make the minimal solution for 3.1.1. Thus, maybe it is easier for you to just submit a new PR with those translations changes not being part of this topic/issue and not to take care of this branch/PR -- I will then remove them from here... As you wish...

File: locale/fr_FR/submission.xml (hunk 9 above)
Comment (forgive38) [reply locale/fr_FR/submission.xml]: @asmecher : why do you think this is not related to this PR ?

Comment (bozana) [reply locale/fr_FR/submission.xml]: think because this changes already existing traslations and we would need an translator (manager) to review that... but, if I am wrong, Alec can say...

**Comment by forgive38** [comment]
@bozana : should I understand that you would prefer a separate PR for update of locale fr_FR ?

File: locale/fr_FR/submission.xml (hunk 9 above)
Comment (bozana) [reply locale/fr_FR/submission.xml]: Hmmm... I saw one more comment in my e-mail notifications but now I cannot see it here :-P @forgive38, yes, those translation changes (especially without changing the en_US texts) should go into a separate PR -- that means that the translation is not quite right or should be improved or ... , right? This would be a topic/PR for it self -- translation improvement... -- and we would need to consult the translator who is responsible for and maintaining that translation to take a look...

**Comment by forgive38** [comment]
@bozana : thank you for the work done

File: locale/fr_FR/submission.xml (hunk 9 above)
Comment (forgive38) [reply locale/fr_FR/submission.xml]: ok I see
I will do a PR to locale update

**Comment by bozana** [comment]
@forgive38, THANKS A LOT to you! I used your PR as basis (it is in the commit history), then made the changes discussed here and created the new PRs, that would minimally solve the current problems:
pkp-lib master: #3399
ui-library master: pkp/ui-library#10
ojs master: pkp/ojs#1840
As you have probably seen, the new/right solution for revision upload notifications is discussed here and will come with the next release: https://github.com/pkp/pkp-lib/issues/2665

Thus, I will close this issue now and am happy to see your new PR with the proposed translation changes...

THANKS A LOT again!
